File: algorithms/neetcode/algorithms/algorithms/src/main/java/com/mycompany/trees/BalancedBinaryTree.java
package com.mycompany.trees;

import javafx.util.Pair;

public class BalancedBinaryTree {

    public static void main(String... args) {

    }

    private static Pair<Boolean, Integer> dfs(TreeNode root) {
        if(root == null) {
            return new Pair<Boolean, Integer>(true, 0);
        }

        Pair<Boolean, Integer> left = dfs(root.left);
        Pair<Boolean, Integer> right = dfs(root.right);

        boolean balanced = left.getKey() &&
                right.getKey() &&
                (Math.abs(left.getValue() - right.getValue()) <= 1);

        return new Pair<Boolean, Integer>(
                balanced,
                1 + Math.max(left.getValue(), right.getValue())
        );

    }

    public static boolean isBalanced(TreeNode root) {
        return dfs(root).getKey();
    }

}

File: algorithms/neetcode/trees/balanced-binary-tree.py

from typing import Optional

import utils

class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.right = right
        self.left = left

class S:

    def isBalanced(self, root: Optional[TreeNode]) -> bool:

        def dfs(root):
            if not root: return [True, 0]

            left, right = dfs(root.left), dfs(root.right)
            balanced = (left[0] and right[0] and
                        abs(left[1] - right[1]) <= 1)
            return [balanced, 1 + max(left[1], right[1])]

        return dfs(root)[0]

array = [1,2,2,3,3,None,None,4,4]

root = utils.Utils().list2tree(array)

print(root)

print(S().isBalanced(root))
